File: moderate/66_pascals_triangle.py
import sys

def gen_pascals_triangle():
    row = [1]
    while True:
        yield row
        row = [0] + row + [0]
        row = [row[i]+row[i+1] for i in range(len(row)-1)]

def pascals_triangle(depth):
    g = gen_pascals_triangle()
    return [next(g) for i in range(depth)]

# Precalculate the triangle to depth 20, the max size that we will be asked for.
triangle = [str(n) for row in pascals_triangle(20) for n in row]

for line in sys.stdin:
    n = int(line)
    print(' '.join(triangle[:(n*(n+1)//2)]))

File: moderate/66_pascals_triangle.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define LINE_LEN  1024
#define MAX_DEPTH 20
#define MAX_NUMBERS (MAX_DEPTH * (MAX_DEPTH + 1) / 2)

void get_pascals_triangle(int depth, int triangle[]) {
    /* Generate Pascal's triangle sequence from depth 1 to `depth`.
     * `triangle` will contain the triangle in row major form.
     * It is assumed that `triangle` is of adequate size which can be
     * calculated using n(n+1)/2 where n = depth.
     */
    int row[MAX_DEPTH+2] = {0};
    row[1] = 1;
    for (int d=1; d<=depth; d++) {
        memcpy(triangle, row+1, sizeof(int)*d);
        triangle += d;

        int tmp_row[MAX_DEPTH+2] = {0};
        for (int i=0; i<=d; i++) {
            tmp_row[i+1] = row[i] + row[i+1];
        }
        memcpy(row, tmp_row, sizeof(int)*(d+2));
    }
}

int main(int argc, const char * argv[]) {
    FILE *file = fopen(argv[1], "r");
    char line[LINE_LEN];
    int triangle[MAX_NUMBERS];

    /* Precalculate the triangle to depth 20, the maximum size that we will be
       asked for. */
    get_pascals_triangle(MAX_DEPTH, triangle);

    while (fgets(line, LINE_LEN, file)) {
        int depth = atoi(line);
        for (int i=0; i<(depth*(depth+1)/2); i++)
            printf("%d ", triangle[i]);
        printf("\n");
    }

    return 0;
}
